src/shared/filesystem.ts

```diff
@@ -6,102 +6,50 @@
 'use strict'
 
 import * as fs from 'fs'
+import * as os from 'os'
+import * as path from 'path'
+import { promisify } from 'util'
 
-export async function accessAsync(path: string | Buffer): Promise<void> {
-    await new Promise<void>((resolve, reject) => fs.access(path, err => {
-        if (!err) {
-            resolve()
-        } else {
-            reject(err)
-        }
-    }))
+const access = promisify(fs.access)
+
+export type PathLike = fs.PathLike
+
+export const accessAsync = async (pathLike: fs.PathLike): Promise<void> => {
+    return access(pathLike)
 }
 
-export async function mkdirAsync(
-    path: fs.PathLike,
-    options?: number | string | fs.MakeDirectoryOptions | undefined | null
-): Promise<void> {
-    await new Promise<void>((resolve, reject) => fs.mkdir(path, options, err => {
-        if (!err) {
-            resolve()
-        } else {
-            reject(err)
-        }
-    }))
+export const mkdirAsync = promisify(fs.mkdir)
+
+const mkdtemp = promisify(fs.mkdtemp)
+
+export const getTempDirPath = (prefix: string = 'vsctk') => {
+    return path.join(
+        os.type() === 'Darwin' ? '/tmp' : os.tmpdir(),
+        prefix || 'vsctk'
+    )
 }
 
-export async function mkdtempAsync(prefix: string): Promise<string> {
-    return await new Promise<string>((resolve, reject) => {
-        fs.mkdtemp(prefix, (err, folder) => {
-            if (!err) {
-                resolve(folder)
-            } else {
-                reject(err)
-            }
-        })
-    })
+export const  mkdtempAsync = async (prefix?: string) => {
+    return mkdtemp(getTempDirPath(prefix))
 }
 
-export async function readdirAsync(
-    path: string | Buffer,
-    options?: {
-        encoding: BufferEncoding | null
-        withFileTypes?: false
-    } | BufferEncoding | undefined | null
-): Promise<string[]> {
-    return await new Promise<string[]>((resolve, reject) => {
-        fs.readdir(path, options, (err, files) => {
-            if (!err) {
-                resolve(files)
-            } else {
-                reject(err)
-            }
-        })
-    })
+const readdir = promisify(fs.readdir)
+export const readdirAsync = async (pathLike: fs.PathLike): Promise<string[]> => {
```

Comment:
Do `readdir` and `readdirAsync` (ditto for the other promisified functions) do the same thing? If so, we should refactor all the code to use one of these and remove one of these.

Reply:
Good question. I did have simply: `export const readdirAsync = promisify(fs.readdir)`

The two are different because it is easier to understand and implement 

```pathLike: fs.PathLike): Promise<string[]>``` 

than the union typed equivalent of: 
```
    /**
     * Asynchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    function readdir(path: PathLike, options: { encoding: "buffer"; withFileTypes?: false } | "buffer", callback: (err: NodeJS.ErrnoException, files: Buffer[]) => void): void;

    /**
     * Asynchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    function readdir(
        path: PathLike,
        options: { encoding?: string | null; withFileTypes?: false } | string | undefined | null,
        callback: (err: NodeJS.ErrnoException, files: string[] | Buffer[]) => void,
    ): void;

    /**
     * Asynchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     */
    function readdir(path: PathLike, callback: (err: NodeJS.ErrnoException, files: string[]) => void): void;

    /**
     * Asynchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options If called with `withFileTypes: true` the result data will be an array of Dirent.
     */
    function readdir(path: PathLike, options: { withFileTypes: true }, callback: (err: NodeJS.ErrnoException, files: Dirent[]) => void): void;
``` 
We aren't using those options so we are truly exposing a simpler interface.

I could remove the non-exported constants and do something like this:
```
export const readFileAsync = async (pathLike: fs.PathLike, encoding?: string): Promise<string | Buffer> => {
    promisify(fs.readdir)(pathLike)
}
```
but that is strange to look at and potentially a tiny bit less efficient since `promisify` would get called every time `readFileAsync` is rather than just once during module initialization.